<a href="https://colab.research.google.com/github/apollonov-evgeniy/2/blob/main/proj14_%D0%B8%D0%B7%D0%B2%D0%BB%D0%B5%D1%87%D0%B5%D0%BD%D0%B8%D0%B5_%D0%B4%D0%B0%D0%BD%D0%BD%D1%8B%D1%85_%D0%B0%D0%BD%D0%B0%D0%BB%D0%B8%D0%B7_%D0%B1%D0%B8%D0%BB%D0%B5%D1%82%D0%BE%D0%B2_%D0%B0%D0%B2%D0%B8%D0%B0%D0%BA%D0%BE%D0%BC%D0%BF%D0%B0%D0%BD%D0%B8%D0%B8_SQL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Вы аналитик российской авиакомпании F9, выполняющей внутренние пассажирские перевозки. Важно понять предпочтения пользователей, покупающих билеты на разные направления.

Вам предстоит изучить базу данных и проанализировать спрос пассажиров на рейсы в города, где проходят крупнейшие культурные фестивали.


<div class="alert alert-info">
Привет! Меня зовут Никита Мишин и я буду твоим ревьюером по этому проекты.
Для простоты предлагаю общение на 'ты'. Буду предполагать, что ты не против:) 
Если более предпочтительно обращение на 'Вы', пиши, не стесняйся.
Также если будут возникать вопросы, аналогично, пиши:)

Предлагаю работать в известном тебе итеративном формате.
Итерация состоит в моей проверке твоего решения. 
После решения могут остаться какие-то недочеты, которые я попрошу тебя устранить, ты их исправляешь и я проверяю твои решения.
    Оставленные мною комментарии могут быть разного вида:
   
    - зеленый: элегантные решения, которые тебе стоит запомнит и в дальнейшем взять на вооружение:) 
    
    - желтый: сигнал о том, что есть некритичная вещь(не всегда ошибка), что нужно точно поправить в следующей работе, даже желательно в этой (полезно, в первую очередь, для тебя:) ).Также это рекомендации на будущее    

    - красный: недочет, который нужно исправить в этой работе, для того, чтобы она была принята
    
    - синий: полезная информация, доп ресурсы, "вопросы на подумать"

Также попрошу не удалять мои комментарии:) <a class="tocSkip">
</div>

Работа с базой  
Менеджер авиаперевозчика готовит исследование. Он хочет выявить связь между числом полётов, типом самолётов и календарём музыкальных фестивалей. Помогите ему получить необходимые данные.
Вы решите задачу в несколько шагов, двигаясь от простых запросов к более изощрённым.  
Описание данных  
В вашем распоряжении база данных об авиаперевозках.  
Таблица airports — информация об аэропортах:  
airport_code — трёхбуквенный код аэропорта  
airport_name — название аэропорта  
city — город  
timezone — временная зона  
Таблица aircrafts — информация о самолётах:  
aircraft_code — код модели самолёта  
model — модель самолёта  
range — дальность полёта  
Таблица tickets — информация о билетах:  
ticket_no — уникальный номер билета  
passenger_id — персональный идентификатор пассажира  
passenger_name — имя и фамилия пассажира  
Таблица flights — информация о рейсах:  
flight_id — уникальный идентификатор рейса  
departure_airport — аэропорт вылета  
departure_time — дата и время вылета  
arrival_airport — аэропорт прилёта  
arrival_time — дата и время прилёта  
aircraft_code — id самолёта  
Таблица ticket_flights — стыковая таблица «рейсы-билеты»  
ticket_no — номер билета  
flight_id — идентификатор рейса  
Таблица festivals — информация о фестивалях  
festival_id — уникальный номер фестиваля  
festival_date — дата проведения фестиваля  
festival_city — город проведения фестиваля  
festival_name — название фестиваля  
Схема таблиц  
image  
Пояснение: В базе данных нет прямой связи между таблицами airports и festivals, а также festivals и flights. Но вы можете писать JOIN и связывать эти таблицы по городу проведения фестиваля (festival_city) и городу аэропорта (city). Потребуется некоторое преобразование дат в flights, и тогда эту таблицу также можно будет связать по дате проведения фестиваля (festival_date) в запросах с JOIN.  

1.  
Найдите количество рейсов на каждой модели самолёта с вылетом в сентябре 2018 года. Назовите получившееся поле flights_amount и выведите его. Также напечатайте на экране поле model. Столбцы в таблице должны быть выведены в следующем порядке:  
model  
flights_amount  
Подсказка  
Соедините таблицы aircrafts и flights. Примените агрегирующие функции с группировкой и не забудьте написать условие.  

SELECT   
        model,  
        count(model) AS flights_amount  
FROM flights  
JOIN aircrafts ON aircrafts.aircraft_code = flights.aircraft_code  
WHERE (flights.departure_time :: date < '2018-10-01'  
           AND flights.departure_time :: date >= '2018-09-01')  
GROUP BY  
        model 
        
 2.  
Посчитайте отправленные в сентябре рейсы самолётов следующих моделей:  
Boeing,  
Airbus,  
другие (“other”).  
Назовите получившуюся переменную flights_amount и выведите её значение на экран.  
Подсказка  
Соедините таблицы aircrafts и flights. Примените конструкцию CASE для группировки моделей.     
SELECT   
    COUNT(flights.flight_id) AS flights_amount,  
    CASE WHEN aircrafts.model LIKE '%Boeing%' THEN 'Boeing'  
         WHEN aircrafts.model LIKE '%Airbus%' THEN 'Airbus'  
         ELSE 'other' END AS type_aircraft  
FROM aircrafts  
INNER JOIN flights ON aircrafts.aircraft_code = flights.aircraft_code  
WHERE (flights.departure_time :: date < '2018-10-01'  
       AND flights.departure_time :: date >= '2018-09-01')  
GROUP BY  
    type_aircraft  
    
3.  
Посчитайте среднее количество прибывающих рейсов в день для каждого города за август 2018 года. Назовите получившееся поле average_flights, вместе с ним выведите столбец city.  
Выводите столбцы в таком порядке:  
city,  
average_flights.  
Подсказка  
Соедините таблицы flights и airports. Посчитайте количество рейсов для каждого города и найдите среднее.  

SELECT subq.city,  
       avg(subq.num_flights) AS average_flights  
FROM  
    (SELECT airports.city AS city,  
            count(flights.flight_id) AS num_flights,  
            extract(DAY 
                    FROM flights.arrival_time :: date) AS DAY  
     FROM flights  
     JOIN airports ON airports.airport_code = flights.arrival_airport  
     WHERE (flights.arrival_time :: date < '2018-09-01'  
            AND flights.arrival_time :: date >= '2018-08-01')  
     GROUP BY airports.city,  
              extract(DAY  
                      FROM flights.arrival_time :: date)) AS subq  
GROUP BY subq.city  

4.  
Установите фестивали, которые проходили с 23 июля по 30 сентября 2018 года в Москве, и номер недели, в которую они проходили. Выведите название фестиваля festival_name и номер недели festival_week.  
Подсказка  
Возьмите таблицу festivals.  

SELECT festival_name AS festival_name,  
       extract(WEEK  
               FROM festival_date) AS festival_week  
FROM festivals  
WHERE festival_city = 'Москва'  
    AND festival_date < '2018-09-30'  
    AND festival_date > '2018-07-23'  
5.  
Для каждой недели с 23 июля по 30 сентября 2018 года посчитайте билеты, которые купили на рейсы в Москву (номер недели week_number и количество билетов ticket_amount). Получите таблицу:  
с количеством купленных за неделю билетов;  
отметкой, проходил ли в эту неделю фестиваль;  
название фестиваля festival_name;  
номер недели week_number.  
Выводите столбцы в таком порядке: - week_number - ticket_amount - festival_week - festival_name  
Подсказка  
Дополните запрос из предыдущего задания.   

SELECT   
	T.week_number,  
	T.ticket_amount,  
	T.festival_week,  
	T.festival_name  
FROM (  
(  
  SELECT EXTRACT(week from flights.arrival_time::date) as week_number,  
          COUNT (ticket_flights.ticket_no) as ticket_amount  
  FROM flights  
  JOIN airports on airports.airport_code = flights.arrival_airport  
  JOIN ticket_flights on ticket_flights.flight_id = flights.flight_id  
  WHERE airports.city = 'Москва' and  
        CAST(flights.arrival_time AS date) BETWEEN '2018-07-23' AND '2018-09-30'  
  GROUP BY week_number  
) t1  
LEFT JOIN   
(SELECT 	  	
		festival_name,	  
		EXTRACT (week FROM festivals.festival_date) AS festival_week  
	FROM   
		festivals  
	WHERE  
		festival_city = 'Москва'  
	  AND CAST(festivals.festival_date AS date) BETWEEN '2018-07-23' AND '2018-09-30'  
) t2   
ON   
	t1.week_number = t2.festival_week  
) AS T  

**Аналитика средствами Python**

У вас есть файлы, в которых содержатся результаты запросов из предыдущих заданий.

**query_1.csv** — результат первого запроса. В нём содержится такая информация:

- *model* — **модель самолета;
- *flights_amount* — количество рейсов для каждой модели самолетов *model* в сентябре 2018 года.

**query_3.csv** — результат третьего запроса. В нём содержится такая информация:

- *city* — город;
- *average_flights* — среднее количество рейсов, прибывающих в город (*city*) за день в сентябре 2018 года.

 Для этих двух наборов данных нужно:

- импортировать файлы;
- изучить данные в них;
- проверить типы данных на корректность;
- выбрать топ-10 городов по количеству рейсов;
- построить графики: модели самолётов и количество рейсов, города и количество рейсов, топ-10 городов и количество рейсов;
- сделать выводы по каждому из графиков, пояснить результат.


# Аналитика

In [ ]:
!pip install cufflinks plotly -U

Defaulting to user installation because normal site-packages is not writeable
Requirement already up-to-date: cufflinks in /home/jovyan/.local/lib/python3.7/site-packages (0.17.3)
Requirement already up-to-date: plotly in /home/jovyan/.local/lib/python3.7/site-packages (4.14.3)


In [ ]:
import os

import numpy as np
import pandas as pd

import seaborn as sns
import matplotlib.pyplot as plt


import math
from scipy import stats as st

import requests

# импортируем модули plotly для интерактивной визуализации
import plotly.graph_objs as go
import cufflinks as cf
import plotly.express as px
from plotly import __version__
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot 
init_notebook_mode(connected=True)
cf.go_offline()

PATH = '/datasets/'
SEED = 21


<div class="alert alert-success">
<h1>Комментарий ревьюера <a class="tocSkip"></a></h1>
Верное решение о расположении всех импортов в начале тетрадки. Так пользователь сразу поймет, какие нужны библиотеки
</div>

## Исследование данных

### Импорт файлов

In [ ]:
query_1 =( pd.read_csv(os.path.join(PATH, 'query_1.csv'))
          .sort_values(by = 'flights_amount',ascending = False)
          .reset_index(drop= True))
query_3 =( pd.read_csv(os.path.join(PATH, 'query_3.csv'))
          .sort_values(by= 'average_flights', ascending =False)
          .reset_index(drop=True))

### Проверка данных

In [ ]:
print('Модели самолётов и количество рейсов')
print('------------------------------------------')
print(query_1.head(10))
print('------------------------------------------')
print(query_1.info())
print('------------------------------------------')
print(query_1.describe())

Модели самолётов и количество рейсов
------------------------------------------
                 model  flights_amount
0   Cessna 208 Caravan            4557
1   Bombardier CRJ-200            4446
2  Sukhoi SuperJet-100            4185
3      Airbus A321-200             960
4       Boeing 737-300             630
5      Airbus A319-100             607
6       Boeing 767-300             600
7       Boeing 777-300             300
------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 2 columns):
model             8 non-null object
flights_amount    8 non-null int64
dtypes: int64(1), object(1)
memory usage: 256.0+ bytes
None
------------------------------------------
       flights_amount
count        8.000000
mean      2035.625000
std       1965.203947
min        300.000000
25%        605.250000
50%        795.000000
75%       4250.250000
max       4557.000000


In [ ]:
print('Города и количество рейсов, среднее в день')
print('------------------------------------------')
print(query_3.head(10))
print('------------------------------------------')
print(query_3.info())
print('------------------------------------------')
print(query_3.describe())

Города и количество рейсов, среднее в день
------------------------------------------
              city  average_flights
0           Москва       129.774194
1  Санкт-Петербург        31.161290
2      Новосибирск        17.322581
3       Красноярск        11.580645
4     Екатеринбург        11.322581
5   Ростов-на-Дону        10.193548
6            Пермь        10.129032
7           Брянск        10.000000
8             Сочи         9.612903
9        Ульяновск         9.580645
------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101 entries, 0 to 100
Data columns (total 2 columns):
city               101 non-null object
average_flights    101 non-null float64
dtypes: float64(1), object(1)
memory usage: 1.7+ KB
None
------------------------------------------
       average_flights
count       101.000000
mean          5.494189
std          13.119790
min           1.000000
25%           2.000000
50%           3.000000
75%           5.870968
max     


<div class="alert alert-warning">
<h1>Комментарий ревьюера <a class="tocSkip"></a></h1>
Дублируется функциональность для разных датасетов. Стоило завести функцию с данной общей функциональностью
</div>

### проверка на наличие NaN значений

In [ ]:
nan_1 = query_1.isna().sum()

nan_2 = query_3.isna().sum()

print('NaN flights amount:\n', nan_1, '\n\n NaN average_flights:\n', nan_2)

NaN flights amount:
 model             0
flights_amount    0
dtype: int64 

 NaN average_flights:
 city               0
average_flights    0
dtype: int64


### Проверка на наличие дубликатов

In [ ]:
duplicate_1 = query_1.duplicated().sum()

duplicate_2 = query_3.duplicated().sum()

print( 'Duplicated flights amount:', duplicate_1,'\n Duplicated average flights:', duplicate_2)

Duplicated flights amount: 0 
 Duplicated average flights: 0



<div class="alert alert-success">
<h1>Комментарий ревьюера <a class="tocSkip"></a></h1>
Молодец, что не забыл проверить на дубликаты
</div>

### Вывод

Данные корректны, типы соответствуют данным, пропусков и дупликатов нет. Корректировки не нуждаются

## ТОП 10 и построение графиков

### ТОП 10 городов по количеству рейсов

In [ ]:
city_top10 = query_3.head(10)
city_top10

city  average_flights
0           Москва       129.774194
1  Санкт-Петербург        31.161290
2      Новосибирск        17.322581
3       Красноярск        11.580645
4     Екатеринбург        11.322581
5   Ростов-на-Дону        10.193548
6            Пермь        10.129032
7           Брянск        10.000000
8             Сочи         9.612903
9        Ульяновск         9.580645

### График ТОП 10 городов по количеству рейсов

In [ ]:
fig = px.bar(city_top10.sort_values(by='average_flights',ascending=True),
             y= 'city', x= 'average_flights',text='average_flights', 
             title='ТОП 10 городов и количество рейсов, среднее за день', 
             color_discrete_sequence=px.colors.sequential.RdBu)
fig.update_traces(texttemplate='%{text:.2s}', textposition='outside')
fig.update_layout(uniformtext_minsize=8, uniformtext_mode='hide')
fig.show()


<div class="alert alert-success">
<h1>Комментарий ревьюера <a class="tocSkip"></a></h1>
График топ!
</div>

#### Вывод

Из графика наибольшее количество рейсов Москва и Санк-Петербург, как большие хабы для осуществления полетов.  
Дальше идет группа аэропортов Сибири, Урала и средней полосы, средние хабы  
Группа "южного направления", срез данных сентябрь месяц, "бархатный сезон отпусков"

### График- модели самолётов и количество рейсов

In [ ]:

fig = px.pie(query_1, values='flights_amount',
             names='model', title='Модели самолётов и количество рейсов', 
             color_discrete_sequence=px.colors.sequential.RdBu)
fig.update_traces(textposition='inside', textinfo='percent+label')
fig.show()

#### Вывод
Первый три места ханимают следющие модели соответсвено: Cessana 208 Caravan, Bombardier CRJ-20 и Sukhoi Superjet -100.  
По количество полётов первые три места с отрывом отличаются от осталных. Их суммарное порцентное соотношение составляет около 80% от общего количетсво полётов.  
Это местные и региональные самолеты.   
Видно, что у авиакомпании большой объем полетов осуществляется небольшими самолетами


<div class="alert alert-success">
<h1>Комментарий ревьюера <a class="tocSkip"></a></h1>
Молодец, верный анализ
</div>

### График - города и количество рейсов

In [ ]:

fig = px.bar(query_3.sort_values(by='average_flights',ascending=True), 
             x='average_flights', y='city', text='average_flights',
             title='Города и количество рейсов, среднее в день',
            color_discrete_sequence=px.colors.sequential.RdBu, height=2000)
fig.update_traces(texttemplate='%{text:.2s}', textposition='outside')
fig.update_layout(uniformtext_minsize=8, uniformtext_mode='hide')
fig.show()

#### Вывод

Из графика видно, что рейсы распределяются на три группы:  
крупные хабы, средние и небольшие аэропорты.

## Вывод

В данном проекте не хватает возможности напрямую из проекта подключиться к pyspark и вытянуть дополнительные аналитики,
т.к. срезы полученные в виде csv не обладают всей информацией.

1. Срез данных за 1 месяц показал, что модели самолетов: *Cessna 208 Caravan*, *Bombardier CRJ-200*, *Sukhoi SuperJet-100*  чаще летали по сравнению с другими фирмами производителями, такими как Boeing и Airbus.  Стоит отметить, что *Cessna 208 Caravan* является лидером по перелетам, но это однотурбинный самолет расчитанный на небольшие расстояния, и используется  для сообщения с небольшими населенными пунктами, где не рентабельно использовать более крупные самолеты и отсутствуют большие аэродромы. 

2. Количество рейсов вероятнее всего соответствует количеству жителей в городе, что  заметно на  графике по соотношению городов и среднему количеству полетов в неделю. 


В дополнение к текущим заданиям, я бы дополнительно исследовал города и топ самолетов для каждого города, сезональные активности, корреляцию между численностью и количеством полетов, модели и вместительность самолетов - пасажиропоток. 


<div class="alert alert-success">
<h1>Комментарий ревьюера v2 <a class="tocSkip"></a></h1>
Классые графки, четкий и лаконичный анализ, лайк!
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"></a></h2>

-Спасибо за качественно сделанный проект, было приятно проверять.

- Код написан хорошо 

- Соблюдена структура проекта 

- Отдельное спасибо за графики



</div>

# Чек-лист проверки

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Файлы изучены
- [x]  Типы данных проверены на корректность
- [x]  Выбран топ-10 городов по количеству рейсов
- [x]  Построен график «Модели самолётов и количество рейсов»
- [x]  Построен график «Города и количество рейсов»
- [x]  Построен график «Топ-10 городов и количество рейсов»
- [x]  На каждом этапе написаны выводы
- [x]  Написан общий вывод